src/zkp/piaffg.rs

```diff
@@ -596,6 +603,353 @@ mod tests {
+        let bad_transformed_ciphertext_verifier = pk.random_ciphertext(&mut rng);
+        assert_ne!(
+            bad_transformed_ciphertext_verifier,
+            input.transformed_ciphertext_verifier
+        );
+        let bad_input = PiAffgInput::new(
+            input.verifier_setup_params.clone(),
+            input.verifier_encryption_key.clone(),
+            input.prover_encryption_key.clone(),
+            input.original_ciphertext_verifier.clone(),
+            bad_transformed_ciphertext_verifier,
+            input.add_coeff_ciphertext_prover.clone(),
+            input.mult_coeff_exp,
+        );
+        assert!(proof.verify(&bad_input, &(), &mut transcript).is_err());
+
+        // Swap verifier setup parameters with a random [`Ciphertext`]
+        let bad_add_coeff_ciphertext_prover = pk.random_ciphertext(&mut rng);
+        assert_ne!(
+            bad_add_coeff_ciphertext_prover,
+            input.transformed_ciphertext_verifier
+        );
+        let bad_input = PiAffgInput::new(
+            input.verifier_setup_params.clone(),
+            input.verifier_encryption_key.clone(),
+            input.prover_encryption_key.clone(),
+            input.original_ciphertext_verifier.clone(),
+            input.transformed_ciphertext_verifier.clone(),
+            bad_add_coeff_ciphertext_prover,
+            input.mult_coeff_exp,
+        );
+        assert!(proof.verify(&bad_input, &(), &mut transcript).is_err());
+
+        // Swap multi coefficient exponent with a random [`CurvePoint`]
+        let mask = random_plusminus_by_size(&mut rng, ELL);
+        let bad_mult_coeff_exp = CurvePoint::GENERATOR.multiply_by_scalar(&mask)?;
+        assert_ne!(bad_mult_coeff_exp, input.mult_coeff_exp);
+        let bad_input = PiAffgInput::new(
+            input.verifier_setup_params,
+            input.verifier_encryption_key,
+            input.prover_encryption_key,
+            input.original_ciphertext_verifier,
+            input.transformed_ciphertext_verifier,
+            input.add_coeff_ciphertext_prover,
+            bad_mult_coeff_exp,
+        );
+        assert!(proof.verify(&bad_input, &(), &mut transcript).is_err());
+
+        Ok(())
+    }
+
+    #[test]
+    fn piaffg_proof_with_inconsistent_secret_inputs() -> Result<()> {
+        let mut rng = init_testing();
+        let x = random_plusminus_by_size(&mut rng, ELL);
+        let y = random_plusminus_by_size(&mut rng, ELL);
+        let (decryption_key_0, _, _) = DecryptionKey::new(&mut rng).unwrap();
+        let pk0 = decryption_key_0.encryption_key();
+
+        let (decryption_key_1, _, _) = DecryptionKey::new(&mut rng).unwrap();
+        let pk1 = decryption_key_1.encryption_key();
+
+        let X = CurvePoint::GENERATOR.multiply_by_scalar(&x)?;
+        let (Y, rho_y) = pk1
+            .encrypt(&mut rng, &y)
+            .map_err(|_| InternalError::InternalInvariantFailed)?;
+
+        let C = pk0.random_ciphertext(&mut rng);
+
+        let (D, rho) = {
```

Comment:
suggestion: I would recommend using the Rust-canonical names from the `Input` type rather than these one-letter capitals.